expression/builtin_cast_vec.go

```diff
@@ -1407,11 +1407,42 @@ func (b *builtinCastStringAsRealSig) vecEvalReal(input *chunk.Chunk, result *chu
 }
 
 func (b *builtinCastStringAsDecimalSig) vectorized() bool {
-	return false
+	return true
 }
 
 func (b *builtinCastStringAsDecimalSig) vecEvalDecimal(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	if IsBinaryLiteral(b.args[0]) {
+		return b.args[0].VecEvalDecimal(b.ctx, input, result)
+	}
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err = b.args[0].VecEvalString(b.ctx, input, buf); err != nil {
+		return err
+	}
+	result.ResizeDecimal(n, false)
+	result.MergeNulls(buf)
+	res := result.Decimals()
+	stmtCtx := b.ctx.GetSessionVars().StmtCtx
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		dec := new(types.MyDecimal)
```

Comment:
We can move this out of the loop to reduce the memory allocation.

Reply:
I have moved it

Reply:
If I  move this out of the loop, the `res[i] = *des` always use the same memory and it will cause an error.

Reply:
So I move it back

Reply:
I still think it can be moved out of the loop. But you need to initialize the struct to the default value inside the loop. Could you please try this again?

Reply:
Like this? 
```go
	dec := new(types.MyDecimal)
	for i := 0; i < n; i++ {
		if result.IsNull(i) {
			continue
		}
		dec.FromInt(0)
```
There is the same error.
>... obtained types.MyDecimal = types.MyDecimal{digitsInt:1, digitsFrac:18, resultFrac:18, negative:false, wordBuf:[9]int32{0, 59365808, 534838596, 0, 0, 0, 0, 0, 0}}
... expected types.MyDecimal = types.MyDecimal{digitsInt:1, digitsFrac:18, resultFrac:18, negative:false, wordBuf:[9]int32{0, 59365808, 534838596, 500000000, 0, 0, 0, 0, 0}}
.

Reply:
Is the default value not zero?

Reply:
Fine, I think it is ok now.